docs/_docs/02_testing-best-practices/testing-plan-output.md

```diff
@@ -0,0 +1,40 @@
+---
+layout: collection-browser-doc
+title: Testing plan output
+category: testing-best-practices
+excerpt: >-
+  Testing specific changes described in Terraform plan output
+tags: ["testing-best-practices"]
+order: 202
+nav_title: Documentation
+nav_title_link: /docs/
+---
+
+Sometimes you can get the information you need in your tests from a `terraform plan`; you may not need to do an `apply`.  Terratest provides a way to use certain information from a `terraform plan` in your tests.
+
+The `InitAndPlanWithInfo` function will return a `PlanInfo` struct that will contain information about which resources would be changed and what specifically about them would change.
+
+For example, let's say you have some `web_server` resources that you're creating with [`for_each`](https://www.terraform.io/docs/configuration/resources.html#for_each-multiple-resource-instances-defined-by-a-map-or-set-of-strings), and you're using the `for_each` values in the name:
+
+```hcl
+resource "aws_instance" "web_server" {
+  for_each = ["env1", "env2", "env3"]
+  name = "web_server_${each.value}"
+  ...
+}
+
+if you want to assert that the names of `web_server` resources that are definied with a terraform `for_each`, you can test this by making assertions against info in this `PlanInfo` struct:
+
+```go
+plan := terraform.InitAndPlanWithInfo(t, terraformOptions)
+
+for _, rc := range plan.PlannedValues {
+    if rc.Name == "web_server" {
+        expected := fmt.Sprintf("web_server_%s", rc.Index)
+        assert.Equal(rc.Values["name"], expected)
```

Comment:
I think Terraform calls them arguments or parameters?

Reply:
This is where it starts to get hairy.  That `Values` struct member is a direct translation from the JSON Terraform plan output, where each resource object has a `"values"` key which is an object from Terraform resource arguments -> values:

```
{
  "address": "null_resource.test[\"val1\"]",
  "mode": "managed",
  "type": "null_resource",
  "name": "test",
  "index": "val1",
  "provider_name": "null",
  "schema_version": 0,
  "values": {
    "triggers": {
      "some_attribute": "attr-val1"
    }
  }
},
```
One question is if we want to go with the terminology of the plan output format as described [here](https://www.terraform.io/docs/internals/json-format.html#values-representation), or if we want to translate it into what we think would make more sense.

My vote, at least initially, would be to stick with the plan output terminology; I think it could be a slippery slope and get more and more confusing if we start straying too far from it.  But of course I'm totally open to disagreement and implementing this in whatever way folks think is best :)

Reply:
Oh, I gotcha.

My guess is most people will:

1. Run `terraform plan`.
1. Write a test checking what they visually see in the `plan` output.

So my initial instinct would be to focus on that model. In fact, reading the JSON output looks a bit painful, so I'm not sure we want to force everyone to do that...

But, I don't have a good mental model of how the JSON output maps to the human-friendly output, and therefore, how much of a rabbit hole it may be to translate between them. So, use your judgment and document whatever you decide :)

Also, sanity check: how stable is that JSON output likely to be?

Reply:
Cool, I'll start down the road of calling that "values" key "attributes".  If it gets too hairy, we can correct the other way.

The JSON output is likely to be stable enough that this would still be useful work, I think:
```
Note: The output includes a format_version key, which currently has major version zero to indicate that the format is experimental and subject to change. A future version will assign a non-zero major version and make stronger promises about compatibility. We do not anticipate any significant breaking changes to the format before its first major version, however.
```
From https://www.terraform.io/docs/internals/json-format.html#values-representation